### DISCUSSION TITLE
Actions Credits

### DISCUSSION THREAD

**Discussion by eliotwrobson** [PrairieLearn/PrairieLearn#6142] [Dev]
It looks like all of the actions credits in the PL organization have been used for this month, which means that the CI jobs for our course can no longer be run (outside of private forks). This is a significant impediment to our development effort, especially because we have several new student developers joining our team with the start of the semester. Is there no way to purchase new actions credits for the organization? We would be happy to contribute some to help offset the cost of this (I believe there's a way to set a spending limit).

Looking forward, issues related to not having ownership over our own course repo are going to show up more and more as our team and codebase grow. I know that the "bring your own" repo arrangement is on the roadmap but not feasible quite yet. Is there any we could contribute any of our development effort towards this?

**Comment by nwalters512** [comment]
I completely understand how frustrating this must be. Clearly having all repos in a single PrairieLearn-managed org has not scaled well 😢 

> Is there no way to purchase new actions credits for the organization? We would be happy to contribute some to help offset the cost of this (I believe there's a way to set a spending limit).

Unfortunately CS 374 is not the only course making extensive usage of GitHub Actions. Looking at the usage report, `pl-cee498ds` (cc @ctessum) and `pl-abe498` (cc @ziyoi) each use roughly as many minutes as CS 374 does. I wouldn't want 374 to have to shoulder all of the cost. That said, minutes are cheap: I suspect it would take on the order of $30-50 per month to cover the minutes over our free allowance (this is assuming moderately increased usage once the semester starts; it may be lower or higher in practice). If you're willing to pay for those minutes, we'd be happy to send you an invoice at the end of each month.

All of the above will of course become a non-issue once we allow a bring-your-own-repo approach; you could create your own org and handle all billing yourself. And on that note...

> Is there any we could contribute any of our development effort towards this?

We'd love to work with y'all to make progress on this! I have a rough idea of what this will have to entail. If you're interested in working on this, I'd encourage you to open up a new discussion and we can start talking technical details.

**Reply by eliotwrobson** [reply]
> That said, minutes are cheap: I suspect it would take on the order of $30-50 per month to cover the minutes over our free allowance (this is assuming moderately increased usage once the semester starts; it may be lower or higher in practice). If you're willing to pay for those minutes, we'd be happy to send you an invoice at the end of each month.

Yes, the pricing on minutes seems very reasonable. Since I don't technically manage expenses for this development effort, I don't want to commit to this _immediately_ (cc @jeffgerickson), but this is very low compared to other expenses, and is extremely worthwhile given the size of our codebase. I think we'd just want to chat briefly about what our actions usage looks like for the last couple of months (to make sure we can keep things under control), but I don't see any reason why not.

> We'd love to work with y'all to make progress on this! I have a rough idea of what this will have to entail. If you're interested in working on this, I'd encourage you to open up a new discussion and we can start talking technical details.

Absolutely! We're starting to scale pretty fast, so I think it's important to get the ball rolling ASAP. Created a discussion thread here: https://github.com/PrairieLearn/PrairieLearn/discussions/6144

**Reply by jeffgerickson** [reply]
I agree that the cost is reasonable, and I certainly don't have a problem contributing.  But experience strongly suggests that the effort required to push an expense through the university accounting system is inversely proportional to the amount of money involved.  Purchasing is an administrative quagmire.

I worry that the easiest way to make this work is to meet in the alley behind Paradiso on the 13th of every month and hand @mwest1066 a short stack of unmarked ten dollar bills.

**Reply by eliotwrobson** [reply]
^I happen to know someone with an unmarked Venmo account overflowing from 10s of dollars sent to them on their birthday.

In total seriousness, I think we haven't run up against this issue THAT often, the last time all the actions credits got spent was by mistake, and our CI job was improved recently to be significantly faster. If you can add like $20 extra or something for this month, that might be enough to get us to September.

**Reply by mwest1066** [reply]
I fully agree with @jeffgerickson about the horror of trying to get this through university purchasing. He also knows exactly where to find me 😄

**Comment by ctessum** [comment]
Sorry 😅! 

I'm theoretically happy to pay, however I get a mini panic attack just thinking about the reaction from the business office if I tried to submit this as an expense.

**Reply by eliotwrobson** [reply]
Out of curiosity, what are your actions configured to do? Ours are running verification / test suites for our Python backend, but I'm wondering what others have done (and if there's functionality that might be useful for us).

**Reply by ctessum** [reply]
Sure. My class has a lot of notebooks that run in workspaces, and the actions are running tests to ensure that the notebooks are working properly, and making sure that when the answer key is substituted in it gives a 100% score. 

It turns out that almost all of the time in the actions is spent setting up the environment and installing packages, so I guess there should be a way to streamline that, perhaps by starting from the docker image that the workspace uses rather than installing all of the packages from scratch.

**Reply by echuber2** [reply]
I agree that it would make sense to do those integration tests with the workspace/grader images.

**Reply by eliotwrobson** [reply]
Ahh I see, I think similar to our use case. We have a bunch of validation / testing for shared code used between multiple questions.

**Reply by mwest1066** [reply]
This seems like a great use of Actions and something we definitely want to support.

**Comment by trombonekenny** [comment]
I'll tack onto the conversation that it seems like we hit our Action limits not necessary from normal use, but from a few runs that get stuck and take dozens of minutes to fail. If the CI pieces should normally happen in a minute or two, there should be safeguards you can put in (either from Actions or your test suite) to timeout and fail early. I haven't looked into it, but I've seen this in other places where a minutes pool are shared between different repositories....

**Reply by nwalters512** [reply]
Interesting - from spot-checking a few repos, it seems like most of the minutes are instead consumed by lots of short-ish runs instead of a few really long ones. Were you looking at any specific ones that took a long time?

Regardless, I agree with folks being responsible and setting safeguards. Another easy one is to limit which files trigger specific workflows: if a commit just changes an HTML file, there's no need to run e.g. workspace integration tests.

**Reply by trombonekenny** [reply]
In the last week, 242 minutes were used in 37 actions. 13 of those used 5 minutes or more, accounting for 190 of those 242 total minutes. That's what led me to say the outliers are the ones eating up the minutes. Since they are on the thread, that included a 68 minute cee498ds CI.yml single run but also multiple 10-20 minute cs374 python-app.yml runs.

If each of those were capped at 2-4 minutes, we wouldn't be out of global minutes.

**Comment by ctessum** [comment]
@ziyoi and I have turned off the computationally intensive part of our actions for now, and will work on streamlining them to take up less time and potentially turning them back on again next month.
